cli/connect/console.go

```diff
@@ -0,0 +1,399 @@
+			line := pipedInputScanner.Text()
+			console.executor(line)
+		}
+		return nil
+	}
+
+	// get options for Prompt instance
+	options := getPromptOptions(console)
+
+	// create Prompt instance
+	console.prompt = prompt.New(
+		console.executor,
+		console.completer,
+		options...,
+	)
+
+	console.prompt.Run()
+
+	return nil
+}
+
+func (console *Console) Eval(funcBody string, args ...interface{}) (interface{}, error) {
+	return console.evalFunc(console, funcBody, args...)
+}
+
+func loadHistory(console *Console) error {
+	var err error
+
+	homeDir, err := common.GetHomeDir()
+	if err != nil {
+		return fmt.Errorf("Failed to get home directory: %s", err)
+	}
+
+	console.historyFilePath = filepath.Join(homeDir, HistoryFileName)
+
+	console.historyLines, err = common.GetLastNLines(console.historyFilePath, MaxHistoryLines)
+	if err != nil {
+		return fmt.Errorf("Failed to read history from file: %s", err)
+	}
+
+	return nil
+}
+
+func detectProtocolAndReconnectIfRequired(console *Console) error {
+	greeting, err := readGreeting(console.conn)
+	if err != nil {
+		return fmt.Errorf("Failed to read Tarantool greeting: %s", err)
+	}
+
+	switch {
+	case strings.Contains(greeting, "(Lua console)"):
+		log.Debugf("Plain text protocol is detected")
+		console.protocol = PlainTextProtocol
+	case strings.Contains(greeting, "(Binary)"):
+		log.Debugf("Binary protocol is detected")
+		console.protocol = BinaryProtocol
+		if err := binaryConnect(console); err != nil {
+			return fmt.Errorf("Failed to connect to binary port: %s", err)
+		}
+	default:
+		return fmt.Errorf("Unknown protocol: %s", greeting)
+	}
+
+	return nil
+}
+
+func getEvalFunc(console *Console) (EvalFunc, error) {
+	switch {
+	case console.protocol == PlainTextProtocol:
+		return plainTextEval, nil
+	case console.protocol == BinaryProtocol:
+		return binaryEval, nil
+	default:
+		return nil, fmt.Errorf("Unknown protocol: %s", console.protocol)
+	}
+}
+
+func getExecutor(console *Console) (prompt.Executor, error) {
+	var executeFunc func(console *Console, in string) string
+	switch {
+	case console.protocol == PlainTextProtocol:
+		executeFunc = plainTextExecute
+	case console.protocol == BinaryProtocol:
+		executeFunc = binaryExecute
+	default:
+		return nil, fmt.Errorf("Unknown protocol: %s", console.protocol)
+	}
+
+	executor := func(in string) {
+		console.input += in + " "
+
+		if !inputIsCompleted(console.input, console.luaState) {
+			console.livePrefixEnabled = true
+			return
+		}
+
+		handleSetOutput(console, console.input)
+
+		if console.historyFilePath != "" {
+			common.AppendStringToFile(console.historyFilePath, console.input+"\n")
+		}
+
+		data := executeFunc(console, console.input)
+
+		fmt.Printf("%s\n", data)
+
+		console.input = ""
+		console.livePrefixEnabled = false
+	}
+
+	return executor, nil
+}
+
+func inputIsCompleted(input string, luaState *lua.LState) bool {
+	if _, err := luaState.LoadString(input); err == nil || !strings.Contains(err.Error(), "at EOF") {
```

Comment:
`!strings.Contains(err.Error(), "at EOF")`

Please, describe it in comment

Reply:
There are two comments below. Added one more with a link to `console` code.